### ISSUE TITLE
KAVA, your script got me into Lua. Here are some questions I’ve been dying to ask.

### ISSUE THREAD

**Issue by LOVEKAVA** [Kava4/PUBG-LOGITECH-NO-RECOIL#62]
**Before asking my questions, I'd like to briefly introduce myself.**
Hello KAVA, I'm a regular gamer living in Korea -- South Korea, not the one with Kim Jong-un and nukes.
I came across your Lua script by chance and was deeply impressed by its precision and structure, which led me to start studying Lua myself.
I've carefully read through each line of your code and learned a lot from it. I wanted to engage in conversation about some questions and possible improvements, so I created a GitHub account for that purpose.
(My nickname was made just for fun and doesn’t reflect my true feelings, so please don’t worry!)

Here are the questions I’d like to ask:

**1. How did you determine the values in your anti-recoil pattern?**

MoveMouseRelativeFractional(0 * Kava * SensSetting, 10 * Kava * SensSetting)
csm(17)

When designing the anti-recoil array, how were values like the ones above calculated?
Did you use a measurement tool, or were the values adjusted based on trial and error?(=Brute force)
(I’m referring to things like the X and Y movement values, the csm delay, number of repetitions, etc.)

**2. Why did you choose to list values line-by-line instead of using tables and for-loops?**
Structurally, the recoil patterns for each weapon seem like they could be simplified using tables and for-loops.
Was there a particular reason you chose a line-by-line approach instead?

**3. Slight inaccuracy when using a table + for-loop based function**
I tried reconstructing the recoil function using a combination of tables and for-loops.
However, I noticed that the left/right recoil seems slightly off compared to the original version.
I’ve double-checked the logic flow, repetition structure, and values multiple times but couldn’t pinpoint the issue.
Are there any common pitfalls or things I should watch out for in such a setup?

※ I plan to create a separate GitHub issue page for these three items.

**4. Possibility of incorporating my version into your code**
If my table/for-loop version of the code proves to be stable and accurate, would you consider incorporating it into your script?

**5. Preferred communication channel**
Is GitHub the best way to reach you for ongoing communication?
If you have a Discord server or another community (such as mpgh.net), I’d be grateful if I could join.
(Currently, mpgh.net shows a “fbi seized” notice, so it seems inaccessible.)

**6. Willingness to contribute**
While I’m still learning, I’d truly like to support your code development in any way I can.
If there are any areas where I could be of help, I’d be more than happy to contribute.

**7. Do you have any plans to create a pattern that accounts for 30% vertical recoil reduction?**
Currently, the Attachment function only applies the compensator,
so it's a bit inconvenient when using other attachments with vertical recoil reduction effects (e.g. vertical grips).
I would really appreciate it if you could consider adding support for patterns with 30% vertical recoil reduction as well.

Thank you for taking the time to read this.
And thank you again for sharing such an excellent script -- I truly appreciate your work.
I look forward to your reply!

**Comment by Kava4** [comment]
**1. How did you determine the values in your anti-recoil pattern?**

Great question, and thank you for taking the time to dive deep into the code.
The recoil values were not created entirely by hand — I actually used a separate Lua script that recorded mouse movement while firing each weapon continuously in training mode. This allowed me to capture the average vertical and horizontal recoil positions over a full magazine.
However, it's important to understand that PUBG has randomized recoil, especially on the horizontal (X) axis. This means no pattern can ever be 100% accurate every time. Because of this, the values I use are meant to represent a general average or baseline — just enough to stabilize the spray and make it easier to control manually.
The csm(delay) values were manually tweaked to match the weapon’s actual fire rate as closely as possible. It’s not perfect, but it keeps the pattern in sync with the game’s firing rhythm.
So, in short: I used recording tools to get a base pattern, and then applied manual smoothing and tuning to make the result consistent and "human-like".

**2: Why did you choose to list values line-by-line instead of using tables and for-loops?**

That’s a very good observation — and yes, using tables with for-loops would definitely make the code look cleaner and more compact.

I actually tried that approach early on, but ended up sticking to the line-by-line method for a few specific reasons:

Consistency and precision: I found that loops — especially when combined with sleep timers — sometimes introduced slight timing inconsistencies, depending on the environment or mouse script engine. Writing each line out explicitly gave me full control over timing and made the recoil more stable during long sprays.

Debugging and pattern tuning: Having each step written line-by-line made it much easier to fine-tune individual steps — especially when I needed to tweak just one Y value or adjust the delay at a specific point in the pattern.

Flexibility: In some patterns, I vary the delay, add conditional logic (e.g. for attachments), or use different scaling mid-pattern. These things are just easier to handle in a more explicit format.

That said, I do think a table-based version could be viable if implemented carefully — especially with a stable loop function and precise timing. So if you manage to recreate it accurately that way, I’d be very happy to take a look!

**3: Slight inaccuracy when using a table + for-loop based function**

I’ve encountered that same issue before — you’re not imagining it.

When using tables + for-loops for recoil, there are a few subtle pitfalls that can cause slight drift or inaccuracy, especially on the horizontal (X) axis, which is more sensitive and randomized in PUBG. Here are a few things to watch for:

1:Timer drift or cumulative delay: When using a loop with Sleep() or similar delay functions, even tiny inconsistencies in timing can accumulate. A difference of just 1–2 ms per loop cycle adds up over a full magazine. This might cause misalignment over time.

2:Floating-point rounding: If you're doing math inside the loop (like step.x * sensitivity * multiplier), small rounding differences can affect precision. Even a 0.1 offset per step becomes noticeable after 20+ steps.

3:Horizontal randomness: PUBG's X-axis recoil is not consistent across sprays. Your table might have the correct average, but it will never match the real-time pattern perfectly. This makes the original pattern (which may have had manual tweaks for left/right) seem more accurate.

4:Environment differences: If you test a table-based version in a different script engine or polling rate (e.g. Logitech G HUB vs older LGS), that might also impact how tight the loop executes.

In short: it’s very possible to get close with a table + loop, but the original line-by-line method allowed me to apply manual micro-adjustments per step, which helped cancel out small inconsistencies. You might try combining both: use a table, but still apply hand-tuned values per weapon instead of calculating them dynamically.

**4: Possibility of incorporating my version into your code**
Absolutely — I’m always open to improvements and contributions, especially if they enhance stability, flexibility, or readability.

If your table-based version performs well and can match the accuracy of the original patterns (or even improve it in some cases), I’d be happy to review it and possibly integrate it as an alternative mode or even as a default structure.

In fact, modular code using tables and loops could make future pattern updates much easier, especially if we want to support more weapons, attachments, or different recoil modes.

Just keep in mind that I’d want to thoroughly test it across various weapons and setups to make sure it behaves consistently before merging it into the main branch. But yes — if it's stable and clean, let’s do it!

**Comment by Kava4** [comment]
The recoil values in the script were primarily developed through careful trial and error combined with observation rather than relying on a specific measurement tool. PUBG’s recoil is partly randomized, so exact values can’t be precisely captured with simple tools. Here’s how the process generally worked:

Starting Point — Visual Observation and Manual Adjustments:
I began by observing the recoil behavior of each weapon in training mode, trying to estimate how much the mouse needed to move vertically and horizontally to compensate for the gun’s kick.
Since PUBG applies some randomness especially on horizontal recoil, the base values reflect averages or common patterns rather than exact repetitions.

Fine-tuning through Repetition:
Using a script, I repeatedly adjusted the mouse movement values, slightly increasing or decreasing the X and Y offsets to find the “sweet spot” where the recoil seemed most controlled over a full spray.
This iterative tuning helped compensate for minor game-induced variability and input lag.

Why Some X Values Are Small:
Horizontal movements of just 1 unit are small but deliberate compensations to counteract the slight left-right sway seen in recoil patterns, which often isn't symmetrical.

On Delays (csm(17) etc.):
Delay values are approximated to match the weapon’s RPM timing as closely as possible, but 1–2 ms differences are negligible due to script execution speed variability and system timing granularity.

Now, addressing the sub-questions about BERRYLL() pattern:
A. Question about total repeat count

The pattern has 149 entries instead of the expected 150 for 30 shots × 5 steps.
Is there a reason one step is missing?

The missing entry is likely due to either optimization or a small oversight. Sometimes the last step is omitted because the recoil pattern effectively “resets” or the spray is assumed to end. It doesn’t usually affect performance because the difference is minimal.

B. Question about first-shot recoil

The first shot recoil seems similar to the second shot (Y=10 twice). Was this intentional?

This was a deliberate simplification. While the first shot typically has stronger recoil, the difference is often handled by the initial trigger pull behavior or player input. The pattern aims for consistency across the burst rather than simulating every nuance, especially since micro-movements are hard to replicate precisely.

C. Question about delay (csm(17))

Using csm(17) gives 0.085 seconds delay, close to the 0.086 seconds for 700 RPM. Why this choice?

The difference of 1 ms is within acceptable margin of error and won’t noticeably affect recoil compensation. System timing and execution delays are rarely precise enough to require perfect matching at the millisecond level.

D. Reason for X-axis movement values

Why are there small horizontal movements of 1 unit? Were they measured or added intentionally?

These small horizontal values are intentional compensations for the horizontal recoil sway, which is not always symmetrical and can vary slightly. They’re not measured with a strict tool but determined by observation and adjusted by hand to improve accuracy.

**Comment by Kava4** [comment]
**Why did you choose to list values line-by-line instead of using tables and for-loops?
And why might a table + for-loop approach cause slight recoil inaccuracies?**

Original choice: line-by-line values
The original script was developed with explicit line-by-line mouse movement commands mainly for precision and control. This approach allows fine-tuning each individual mouse movement step without abstraction, which is helpful because:

Recoil in PUBG is partly randomized: Small timing and positional variations happen each shot, so the pattern isn’t perfectly predictable.

Explicit commands reduce complexity: Listing each step clearly ensures no unintended logic bugs are introduced by loops or indexing.

Debugging is easier: When a single movement causes issues, you can isolate it immediately in a long list rather than tracking loop iterations and table indices.

Why loops + tables feel slightly off
Using tables and loops to condense the pattern is conceptually cleaner and more scalable, but it introduces some pitfalls:

Indexing errors: It’s easy to get off-by-one errors or mix up the order of moves and repeats, especially if repeat counts are involved.

Timing and delays: If the loop timing or the way delays are implemented differs slightly from the original, the recoil compensation can feel mismatched.

Accumulation rounding: Multiplying or scaling values in loops can cause small floating-point rounding errors that accumulate, shifting the cursor unexpectedly.

Side effects from state variables: Sometimes state (like cumulative recoil offset) isn’t handled identically inside a loop as it is in linear code.

Recommendations to avoid inaccuracies
If you want to use tables + loops, consider:

Strictly validate the order of moves: Make sure the table entries follow the exact sequence of X, Y, repeat counts as in the original.

Match delays precisely: Use the same delay function and ensure no extra waits or skipped frames occur inside the loop.

Avoid floating-point accumulation: Use integer increments or carefully round values to prevent drift.

Test with small patterns first: Debug by running short, simple recoil patterns to verify correctness before scaling up.

Log or print each step during loop execution: This helps compare actual behavior vs. expected steps.

Conclusion
The line-by-line style was chosen for robustness and to minimize subtle timing or rounding errors in a highly sensitive context like recoil compensation.
That said, well-structured table + loop approaches can definitely work with careful implementation and thorough testing.
